File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/support/GQLScalarLocalDateTime.java
package net.nemerosa.ontrack.graphql.support;

import graphql.schema.Coercing;
import graphql.schema.GraphQLScalarType;
import net.nemerosa.ontrack.common.Time;

import java.time.LocalDateTime;

/**
 * JSON scalar type.
 */
public final class GQLScalarLocalDateTime extends GraphQLScalarType {

    public static final GraphQLScalarType INSTANCE = new GQLScalarLocalDateTime();

    private GQLScalarLocalDateTime() {
        super(
                "LocalDateTime",
                "Local Date Time",
                new Coercing() {

                    @Override
                    public String serialize(Object input) {
                        if (input instanceof LocalDateTime) {
                            return Time.forStorage((LocalDateTime) input);
                        } else if (input instanceof String) {
                            return (String) input;
                        } else {
                            return null;
                        }
                    }

                    @Override
                    public LocalDateTime parseValue(Object input) {
                        if (input instanceof LocalDateTime) {
                            return (LocalDateTime) input;
                        } else if (input instanceof String) {
                            return Time.fromStorage((String) input);
                        } else {
                            return null;
                        }
                    }

                    @Override
                    public Object parseLiteral(Object input) {
                        return parseValue(input);
                    }
                }
        );
    }
}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/support/GQLScalarLocalDateTime.kt
package net.nemerosa.ontrack.graphql.support

import graphql.language.StringValue
import graphql.schema.*
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException

/**
 * JSON scalar type.
 */
class GQLScalarLocalDateTime private constructor() : GraphQLScalarType(
    "LocalDateTime",
    "Local Date Time",
    object : Coercing<LocalDateTime, String?> {

        override fun serialize(dataFetcherResult: Any): String? =
            if (dataFetcherResult is LocalDateTime) {
                dataFetcherResult.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)
            } else {
                throw CoercingSerializeException("Cannot serialize ${dataFetcherResult::class.java} into a string")
            }

        override fun parseValue(input: Any): LocalDateTime =
            when (input) {
                is String -> try {
                    parse(input)
                } catch (ex: DateTimeParseException) {
                    throw CoercingParseValueException("Cannot parse value: $input", ex)
                }
                is LocalDateTime -> input
                else -> throw CoercingParseValueException("Cannot parse value: $input")
            }

        override fun parseLiteral(input: Any): LocalDateTime =
            when (input) {
                is StringValue -> try {
                    parse(input.value)
                } catch (ex: DateTimeParseException) {
                    throw CoercingParseLiteralException("Cannot parse literal: $input", ex)
                }
                else -> throw CoercingParseLiteralException("Cannot parse literal: $input")
            }

        private fun parse(input: String): LocalDateTime =
            // Tries first with local date time
            try {
                LocalDateTime.parse(input, DateTimeFormatter.ISO_LOCAL_DATE_TIME)
            } catch (ignored: DateTimeParseException) {
                // ... with a time zone
                LocalDateTime.ofInstant(Instant.parse(input), ZoneOffset.UTC)
            }
    }
) {

    companion object {
        val INSTANCE = GQLScalarLocalDateTime()
    }

}
